### ISSUE TITLE
tf 2.15.0.post1 complains about tf.keras.optimizers.Adam to be not trackable

### ISSUE THREAD

**Issue by Binjian** [tensorflow/tensorflow#62723]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

v2.15.0-2-g0b15fdfcb3f 2.15.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

Linux Ubuntu 22.04

### Python version

3.11.7

### CUDA/cuDNN version

12.2.140

### GPU model and memory

NVidia 3080 with 10240MB

### Current behavior?

1. define  a model and a Keras Adam optimizer
```python
import tensorflow as tf
import keras

inputs = keras.Input(shape=(37,))
x = keras.layers.Dense(32, activation="relu")(inputs)
outputs = keras.layers.Dense(5, activation="softmax")(x)
model = keras.Model(inputs=inputs, outputs=outputs)

optimizer = tf.keras.optimizers.Adam(0.001) 
```

2. create a checkpoint
```python
ckpt = tf.train.Checkpoint(
            step=tf.Variable(1, name="step"),
            optimizer=optimizer,
            net=model,
        )
```

3. Output error
```python
ValueError: `Checkpoint` was expecting optimizer to be a trackable object (an object derived from `Trackable`), got <keras.src.optimizers.adam.Adam object at 0x7f836446dad0>. If you believe this object should be trackable (i.e. it is part of the TensorFlow Python API and manages state), please open an issue.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import keras

inputs = keras.Input(shape=(37,))
x = keras.layers.Dense(32, activation="relu")(inputs)
outputs = keras.layers.Dense(5, activation="softmax")(x)
model = keras.Model(inputs=inputs, outputs=outputs)

optimizer = keras.optimizers.Adam(0.001) 

ckpt_actor = tf.train.Checkpoint(
            step=tf.Variable(1, name="step"),
            optimizer=optimizer,
            net=model
        )
```

### Relevant log output

```shell
{
	"name": "ValueError",
	"message": "`Checkpoint` was expecting optimizer to be a trackable object (an object derived from `Trackable`), got <keras.src.optimizers.adam.Adam object at 0x7f836446dad0>. If you believe this object should be trackable (i.e. it is part of the TensorFlow Python API and manages state), please open an issue.",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[10], line 1
----> 1 ckpt_actor = tf.train.Checkpoint(
      2             step=tf.Variable(1, name=\"step\"),
      3             optimizer=optimizer,
      4             net=model
      5         )

File /dpt/.pyenv/versions/miniconda3-3.11-23.11.0-2/envs/meos-nbd/lib/python3.11/site-packages/tensorflow/python/checkpoint/checkpoint.py:2200, in Checkpoint.__init__(self, root, **kwargs)
   2198 if isinstance(converted_v, weakref.ref):
   2199   converted_v = converted_v()
-> 2200 _assert_trackable(converted_v, k)
   2202 if root:
   2203   # Make sure that root doesn't already have dependencies with these names
   2204   child = trackable_root._lookup_dependency(k)

File /dpt/.pyenv/versions/miniconda3-3.11-23.11.0-2/envs/meos-nbd/lib/python3.11/site-packages/tensorflow/python/checkpoint/checkpoint.py:1548, in _assert_trackable(obj, name)
   1545 def _assert_trackable(obj, name):
   1546   if not isinstance(
   1547       obj, (base.Trackable, def_function.Function)):
-> 1548     raise ValueError(
   1549         f\"`Checkpoint` was expecting {name} to be a trackable object (an \"
   1550         f\"object derived from `Trackable`), got {obj}. If you believe this \"
   1551         \"object should be trackable (i.e. it is part of the \"
   1552         \"TensorFlow Python API and manages state), please open an issue.\")

ValueError: `Checkpoint` was expecting optimizer to be a trackable object (an object derived from `Trackable`), got <keras.src.optimizers.adam.Adam object at 0x7f836446dad0>. If you believe this object should be trackable (i.e. it is part of the TensorFlow Python API and manages state), please open an issue."
}
```

**Comment by Venkat6871** [comment]
Hi **@Binjian** ,
I was able to reproduce this issue on Colab using TF v2.15, TF-nightly, Please find the [gist](https://colab.research.google.com/gist/Venkat6871/43f392f1cd7ab2c28cca41dceceafee8/62723_2-15-v.ipynb) here for reference.

Thank you!

**Comment by Binjian** [comment]
@Venkat6871 ,

thank you for the response! Based on you label "keras", I was able to track down an issue in keras back in 2021:

https://github.com/keras-team/keras/issues/14776

It seems to relate to the way Keras implemented its trackable interface of optimizers. I couldn't find the documentation for the typing of the trackable interface.

**Comment by Venkat6871** [comment]
Hi **@Binjian** ,
Could you close this issue as it is fixed in latest version of TensorFlow. Please feel free to reopen the issue if you still have a concern. Here i am providing [gist](https://colab.sandbox.google.com/gist/Venkat6871/1ee849fceebd976190d4396573799479/62723_2-16-1-v.ipynb) for your reference.

Thank you!

**Comment by Binjian** [comment]
Many thanks for the reminder. My bad.

**Comment by XXZhou25** [comment]
> Hello, this bug has been fixed in Keras 3.1.1. Optimizers are now trackable. TensorFlow 2.16 uses the latest version of Keras. So if you use TensorFlow 2.16, you should find that the bug is fixed.

Hi, I encountered exactly the same issue for tensorflow 2.16.1. Here's a simplified version of the code that reproduces the problem:
`
import tensorflow as tf
optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)
tf.train.Checkpoint(optimizer=optimizer)
`
Could you please confirm if TensorFlow 2.16.1 has addressed this bug? If not, do you have any suggestions or workarounds?

Thank you for your assistance.

**Comment by XXZhou25** [comment]
> > Hello, this bug has been fixed in Keras 3.1.1. Optimizers are now trackable. TensorFlow 2.16 uses the latest version of Keras. So if you use TensorFlow 2.16, you should find that the bug is fixed.
> 
> Hi, I encountered exactly the same issue for tensorflow 2.16.1. Here's a simplified version of the code that reproduces the problem: `import tensorflow as tf optimizer = tf.keras.optimizers.SGD(learning_rate=0.01) tf.train.Checkpoint(optimizer=optimizer)` Could you please confirm if TensorFlow 2.16.1 has addressed this bug? If not, do you have any suggestions or workarounds?
> 
> Thank you for your assistance.

I think I found the issue.
I am using tensorflow 2.16.1 gpu-jupyter docker image, and the keras inside was the old version `3.0.5` with bug. So I have to upgrade it to the newer version `3.1.1` and the error was fixed.
